ferrocene/doc/internal-procedures/src/working-with-the-ci.rst

```diff
@@ -0,0 +1,250 @@
+methods:
+
+.. code-block:: bash
+
+   curl -LsSf https://astral.sh/uv/install.sh | sh
+
+.. code-block:: powershell
+
+   powershell -c "irm https://astral.sh/uv/install.ps1 | iex"
+
+Follow the onscreen instructions post-install to ensure ``uv`` is present on your
+``$PATH``.
+
+Set up the virtual environment:
+
+.. code-block:: bash
+
+   uv venv
+   source .venv/bin/activate
+   uv pip sync requirements.txt
+
+You need to source the virtual enviroment each time you wish to use it:
+
+.. code-block:: bash
+
+   source .venv/bin/activate
+
+
+Using the CI ``config.toml``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+To create the ``config.toml`` used by the CI, set ``FERROCENE_HOST`` to your host triple,
+then run the ``configure.sh``:
+
+.. warning::
+
+   This will overwrite your existing `config.toml`, consider backing it up.
+
+.. code-block:: bash
+
+   CI=true FERROCENE_HOST=aarch64-unknown-linux-gnu ./ferrocene/ci/configure.sh 
+
+
+Using the CI Docker images
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Docker images used in CI can be found in ``ferrocene/ci/docker-images``. Each
+image documents how to build it at the top of the ``Dockerfile``.
+
+To build the most common image, ``ubuntu-20``:
+
+.. code-block:: bash
+
+   cd $FERROCENE_REPO
+   docker build -t ubuntu-20 -f ferrocene/ci/docker-images/ubuntu-20/Dockerfile .
+
+
+Reproducing CI jobs
+^^^^^^^^^^^^^^^^^^^
+
+Most CI jobs are formatted similar to this:
+
+.. code-block:: YAML
+   
+  x86_64-linux-dist-targets:
+    executor: docker-ubuntu-20
+    resource_class: large # 4-core
+    environment:
+      FERROCENE_HOST: x86_64-unknown-linux-gnu
+      FERROCENE_TARGETS: << pipeline.parameters.targets--x86_64-unknown-linux-gnu--std-only >>
+      SCRIPT: |
+        ./x.py --stage 2 dist rust-std
+    steps:
+      - ferrocene-job-dist:
+          restore-from-job: x86_64-linux-build
+
+Jobs can only be reproduced on the host triple specified in ``FERROCENE_HOST``.
+Ensure your host is correct.
+
+If ``FERROCENE_TARGETS`` is different than your host triple,
+update your ``config.toml``'s ``[build]`` section's ``host`` and ``target`` to
+reflect that.
+
+For Mac or Windows jobs, you can coarsely reproduce the CI by running the
+lines the ``SCRIPT``, line by line if desired.
+
+For Linux jobs, enter the Docker container specified by the ``executor`` line:
+
+.. code-block:: bash
+
+   docker run --rm -ti -w /ferrocene \
+      --mount "type=bind,src=$(pwd),dst=/ferrocene" \
+      ubuntu-20 bash
+
+.. note::
+
+   If you wish to preserve your ``build/`` artifacts, it may make sense to
+   re-clone the Ferrocene repository inside the container.
+
+Inside the container, run ``./x clean`` then run the lines of the ``SCRIPT``
+of the job:
+
+.. code-block:: bash
+
+   ./x clean
+   ./x --stage 2 dist rust-std
+
+
+Making changes to the CI
+------------------------
+
+Effort should be made to avoid tying to tightly to CircleCI, Ferrocene is
```

Comment:
```suggestion
Effort should be made to avoid tying too tightly to CircleCI... Ferrocene is
```

Reply:
I opted for a single period here instead. I'm hesitant to include `...` in writing like this as it implies hesitation or that some content is ellided.

Reply:
alright, was unsure of it anyways... took a chance at suggesting it, to also learn ingleesh